apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/MetricRegistrySerializer.java

```diff
@@ -91,23 +93,121 @@ private static void serializeSamples(Map<String, DoubleSupplier> samples, JsonWr
+                Map.Entry<String, Timer> kv = iterator.next();
+                if (kv.getValue().hasContent()) {
+                    value = kv.getValue();
+                    if (hasSamples) {
+                        jw.writeByte(JsonWriter.COMMA);
+                    }
+                    hasSamples = true;
+                    serializeTimer(kv.getKey(), value, jw);
+                }
+            }
+
+            // serialize rest
+            while (iterator.hasNext()) {
+                Map.Entry<String, Timer> kv = iterator.next();
+                value = kv.getValue();
+                if (value.hasContent()) {
+                    jw.writeByte(JsonWriter.COMMA);
+                    serializeTimer(kv.getKey(), value, jw);
+                }
+            }
+        }
+        return hasSamples;
+    }
+
+    private static void serializeCounters(Map<String, AtomicLong> timers, boolean hasSamples, JsonWriter jw) {
+
+        final int size = timers.size();
+        if (size > 0) {
+            final Iterator<Map.Entry<String, AtomicLong>> iterator = timers.entrySet().iterator();
+
+            // serialize first valid value
+            AtomicLong value = null;
+            while (iterator.hasNext() && value == null) {
+                Map.Entry<String, AtomicLong> kv = iterator.next();
+                if (kv.getValue().get() > 0) {
+                    value = kv.getValue();
+                    if (hasSamples) {
+                        jw.writeByte(JsonWriter.COMMA);
+                    }
+                    serializeCounter(kv.getKey(), value, jw);
+                }
+            }
+
+            // serialize rest
+            while (iterator.hasNext()) {
+                Map.Entry<String, AtomicLong> kv = iterator.next();
+                value = kv.getValue();
+                if (kv.getValue().get() > 0) {
+                    jw.writeByte(JsonWriter.COMMA);
+                    serializeCounter(kv.getKey(), value, jw);
                 }
             }
         }
+    }
+
+    private static void serializeCounter(String key, AtomicLong value, JsonWriter jw) {
+        serializeValueStart(key, "", jw);
+        NumberConverter.serialize(value.get(), jw);
         jw.writeByte(JsonWriter.OBJECT_END);
+        value.set(0);
     }
 
     private static boolean isValid(double value) {
         return !Double.isInfinite(value) && !Double.isNaN(value);
     }
 
-    private static void serializeSample(String key, double value, JsonWriter jw) {
-        jw.writeString(key);
-        jw.writeByte(JsonWriter.SEMI);
-        jw.writeByte(JsonWriter.OBJECT_START);
-        DslJsonSerializer.writeFieldName("value", jw);
+    private static void serializeTimer(String key, Timer timer, JsonWriter jw) {
+        serializeValue(key, ".count", timer.getCount(), jw);
+        jw.writeByte(JsonWriter.COMMA);
+        serializeValue(key, ".sum", timer.getTotalTimeMs(), jw);
+        timer.resetState();
```

Comment:
All these resets seem awkward here, much better fit in `metricSet.onAfterReport()`. Is this done here in order to prevent additional iteration?

Reply:
Yes, that the idea was to prevent another iteration. But it's probably better to trade off in favor of readability here.